### PR TITLE
graphql: utilize a tower layer for more accurate tracking of graphql metrics

### PR DESCRIPTION
**Pull request by bmwill**
## Description 

Implement a Callback tower layer, which is then used in the Graphql Axum server to be able to properly handle metrics. This way we are properly tracking in-flight requests, since we're using an RAII guard which will decrement when dropped.

## Test Plan 

How did you test the new or updated feature?

---
If your changes are not user-facing and do not break anything, you can skip the following section. Otherwise, please briefly describe what has changed under the Release Notes section.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/mysten-network/src/callback/service.rs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    /// [`Layer`]: tower::layer::Layer
+    pub fn layer(make_handler: M) -> CallbackLayer<M>
+    where
+        M: MakeCallbackHandler,
+    {
+        CallbackLayer::new(make_handler)
+    }
+}
+
+impl<S, M> Callback<S, M> {
```
Comment (stefan-mysten) [inline crates/mysten-network/src/callback/service.rs:37]: What is the advantage of having two impl for the same type?

File: crates/sui-graphql-rpc/src/server/builder.rs [hunk 2]
```diff
@@ -297,15 +301,11 @@ pub fn export_schema() -> String {
...
-    State(metrics): State<Metrics>,
     ConnectInfo(addr): ConnectInfo<SocketAddr>,
     schema: axum::Extension<SuiGraphQLSchema>,
     headers: HeaderMap,
     req: GraphQLRequest,
-) -> GraphQLResponse {
-    metrics.request_metrics.inflight_requests.inc();
-    metrics.inc_num_queries();
-    let instant = Instant::now();
+) -> (axum::http::Extensions, GraphQLResponse) {
```
Comment (stefan-mysten) [inline crates/sui-graphql-rpc/src/server/builder.rs:308]: Is this a known pattern? I am trying to understand how this works...

File: crates/mysten-network/src/callback/service.rs (hunk 1 above)
Comment (bmwill) [reply crates/mysten-network/src/callback/service.rs:37]: A Layer type and Service type are two different things. You can look at the tower ecosystem for more info on this.

File: crates/sui-graphql-rpc/src/server/builder.rs (hunk 2 above)
Comment (bmwill) [reply crates/sui-graphql-rpc/src/server/builder.rs:308]: Yeah, look at the Axum docs for response types. The return type just needs to impl IntoResponse. There's a bit of type magic happening

File: crates/mysten-network/src/callback/service.rs (hunk 1 above)
Comment (stefan-mysten) [reply crates/mysten-network/src/callback/service.rs:37]: I am likely misunderstanding things, but line 17 and line 37 look the same to me? And the functions following the impl block on line 37 do not seem to use the M type. How should I interpret these two impl blocks?

Comment (bmwill) [reply crates/mysten-network/src/callback/service.rs:37]: Ah I see what you mean. No there's no reason

Comment (bmwill) [reply crates/mysten-network/src/callback/service.rs]: Fixed in the latest version

**Review by stefan-mysten — APPROVED** [review]
I don't follow very well the callback layer part, a lot of type magic there with which I am not familiar, but the graphql side looks good. Thanks @bmwill

**Pull request merged**